fastlane/fastlane
# [Fastlane.swift] Fix adding groups to `Fastlane Runner` group and custom `upgrade_manifest.json`

### Checklist

### Motivation and Context



This allows `FastlaneRunner` executable to automatically update when a file inside the nested group is changed.
```
FastlaneSwiftRunner
--> Fastlane Runner
-- --> New Group
-- --> Autogenerated API
```

### Description

Read folders structure from local `upgrade_manifest.json`, map to Fastlane's `upgrade_manifest.json`, copy only files appearing in both manifest (default files).

Without this change:
- Adding groups to `Fastlane Runner` group will cause the error
```
loading manifest: /Users/blys/.gems/gems/fastlane-2.210.1/fastlane/swift/upgrade_manifest.json
+-----------+-------------+
|      Lane Context       |
+-----------+-------------+
| LANE_NAME | custom3Lane |
+-----------+-------------+
[10:15:57]: undefined method `each' for nil:NilClass
[10:15:57]: fastlane finished with errors
```
- Adding group to the project `FastlaneSwiftRunner` is only allow at the root level of the project.
```
FastlaneSwiftRunner
--> Fastlane Runner
-- --> Autogenerated API
--> New Group
```
- Adding the group in the above approach does not allow `FastlaneRunner` to automatically update when the file in `New Group` is changed. User MUST re-save a file in the Group `Fastlane Runner` (without change) to allow `FastlaneRunner` executable to change.

With this change:
- Allow adding group under the group `Fastlane Runner`
<img width="251" alt="Screen Shot 2022-11-14 at 11 29 59" src="https://user-images.githubusercontent.com/6356137/201583880-7f29bbbc-05ec-4ce7-955b-fd9de90dc6e9.png">

Modified `upgrade_manifest.json`
```
{"Actions.swift":"Autogenerated API","Fastlane.swift":"Autogenerated API","DeliverfileProtocol.swift":"Autogenerated API","GymfileProtocol.swift":"Autogenerated API","MatchfileProtocol.swift":"Autogenerated API","Plugins.swift":"Autogenerated API","PrecheckfileProtocol.swift":"Autogenerated API","ScanfileProtocol.swift":"Autogenerated API","ScreengrabfileProtocol.swift":"Autogenerated API","SnapshotfileProtocol.swift":"Autogenerated API","LaneFileProtocol.swift":"Fastfile Components","OptionalConfigValue.swift":"Fastfile Components","Atomic.swift":"Networking","ControlCommand.swift":"Networking","RubyCommand.swift":"Networking","RubyCommandable.swift":"Networking","Runner.swift":"Networking","SocketClient.swift":"Networking","SocketClientDelegateProtocol.swift":"Networking","SocketResponse.swift":"Networking","main.swift":"Runner Code","ArgumentProcessor.swift":"Runner Code","RunnerArgument.swift":"Runner Code",
"Constants/Constant.swift":"Constants",
"Constants/Secret.swift":"Constants",
"Helpers/AppStoreAuthentication.swift":"Helpers",
"Helpers/Build.swift":"Helpers",
"Helpers/Distribution.swift":"Helpers",
"Helpers/EnvironmentParser.swift":"Helpers"
}
``` 

- When adding a group to the group `Fastlane Runner`, user need to modify `upgrade_manifest.json` to include the new folders.

- When modifying a file specified in the `upgrade_manifest.json`, `FastlaneRunner` will automatically update the changes.




### Testing Steps

1. Use `fastlane init swift` on a new project.
2. Add new group `Constants` to `Fastlane Runner` group.
3. Modify local `upgrade_manifest.json` to include the new files and group.
4. Add swift files to `Constants` group.
5. Call new swift files class in `fastfile.swift`.
6. Run fastlane swift.
7. Observe changes.
8. Modify files in `Constants` group.
9. Run fastlane swift.
10. Observe change.
